jerry-core/parser/regexp/re-parser.cpp

```diff
@@ -563,102 +637,123 @@ re_parse_next_token (re_parser_ctx_t *parser_ctx_p, /**< RegExp parser context *
+        advance = 0;
+        lit_utf8_iterator_advance (&(parser_ctx_p->iter), 2);
+
+        if (lit_utf8_iterator_is_eos (&(parser_ctx_p->iter)))
+        {
+          ret_value = ecma_raise_syntax_error ("invalid character class");
+          return ret_value;
+        }
+
+        ecma_char_t ch2 = lit_utf8_iterator_read_next (&(parser_ctx_p->iter));
 
         if ((ch2 >= 'A' && ch2 <= 'Z') || (ch2 >= 'a' && ch2 <= 'z'))
         {
-          advance = 3;
           out_token_p->type = RE_TOK_CHAR;
           out_token_p->value = (ch2 % 32);
         }
-        else
-        {
-          ret_value = ecma_raise_syntax_error ("invalid regexp control escape");
-          break;
-        }
       }
       else if (ch1 == 'x'
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 2))
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 3)))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 2))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 3)))
       {
-        advance = 4;
+        advance = 0;
+        lit_utf8_iterator_advance (&(parser_ctx_p->iter), 2);
+
+        ecma_char_t codeunit = LIT_UNICODE_CODE_POINT_NULL;
+        ret_value = re_get_codeunit_from_string (&(parser_ctx_p->iter), &codeunit, true);
+
+        if (ecma_is_completion_value_throw (ret_value))
+        {
+          return ret_value; /* error */
+        }
+
         out_token_p->type = RE_TOK_CHAR;
-        /* FIXME: get unicode char from hex-digits */
-        /* result.value = ...; */
+        out_token_p->value = codeunit;
       }
       else if (ch1 == 'u'
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 2))
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 3))
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 4))
-               && lit_char_is_hex_digit (RE_LOOKUP (parser_ctx_p->current_char_p, 5)))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 2))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 3))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 4))
+               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 5)))
       {
-        advance = 4;
+        advance = 0;
+        lit_utf8_iterator_advance (&(parser_ctx_p->iter), 2);
+
+        ecma_char_t codeunit = LIT_UNICODE_CODE_POINT_NULL;
+        ret_value = re_get_codeunit_from_string (&(parser_ctx_p->iter), &codeunit, false);
+
+        if (ecma_is_completion_value_throw (ret_value))
+        {
+          return ret_value; /* error */
+        }
+
         out_token_p->type = RE_TOK_CHAR;
-        /* FIXME: get unicode char from digits */
-        /* result.value = ...; */
+        out_token_p->value = codeunit;
       }
       else if (ch1 == 'd')
       {
-        advance = 2;
+        advance = 0;
```

Comment:
In the line 625 you can see that advance is set to 2. If set it to 0 there, then there will be a lot more sets to 2 in other cases.